tests/trackingtest.py

```diff
@@ -0,0 +1,404 @@
+                filepath = out_dir / test.filename
+                if not filepath.exists():
+                    rec_meta = api.query_rec(test.rec_id)
+                    if api.save_file(
+                        out_dir / test.filename,
+                        api._download_signed(rec_meta["downloadRawJWT"]),
+                    ):
+                        logging.info("Saved %s", filepath)
+        self.results = []
+
+    def run_tests(self, args):
+        out_dir = Path(self.test_config.clip_dir)
+
+        for test in self.test_config.recording_tests:
+            filepath = out_dir / test.filename
+            if not filepath.exists():
+                logging.info(
+                    " %s not found, add cmd args user and password to download this file from the cacophony server",
+                    filepath,
+                )
+                continue
+            logging.info("testing %s ", test.filename)
+            clip, predictions = self.clip_classifier.classify_file(filepath)
+            rec_match = self.compare_output(clip, predictions, test)
+            if self.clip_classifier.previewer:
+                mpeg_filename = filepath.with_suffix(".mp4")
+                logging.info("Exporting preview to '%s'", mpeg_filename)
+                self.clip_classifier.previewer.export_clip_preview(
+                    mpeg_filename, clip, predictions
+                )
+            self.results.append(rec_match)
+
+    def write_results(self, filename):
+        with open(filename, "w") as f:
+            for res in self.results:
+                res.write_results(f)
+            f.write("Config\n")
+            json.dump(self.classifier_config, f, indent=2, default=convert_to_dict)
+
+    def print_summary(self):
+        print("===== SUMMARY =====")
+        total_summary = None
+        total_tracks = 0
+        for result in self.results:
+            total_tracks += result.number_tracks
+            summary = result.print_summary()
+            if total_summary is None:
+                total_summary = summary
+            else:
+                for key, value in summary["classify"].items():
+                    total_summary["classify"][key] += value
+
+                for key, value in summary["tracking"].items():
+                    total_summary["tracking"][key] += value
+        if total_summary is None:
+            return
+        classified_correct = total_summary.get("classify", {}).get("correct", 0)
+        tracked_well = total_summary.get("tracking", {}).get("better", 0)
+        tracked_well += total_summary.get("tracking", {}).get("same", 0)
+        unmatched_tracks = total_summary.get("tracking", {}).get("unmatched_tracks", 0)
+        unmatched_tests = total_summary.get("tracking", {}).get("unmatched_tests", 0)
+
+        classified_per = round(100.0 * classified_correct / total_tracks)
+        tracked_per = round(100.0 * tracked_well / total_tracks)
```

Comment:
What if there are extra tracks or missing tracks.   Is it possible to get 100% tracked well?  

Eg If there is one missing track but the existing tracks are perfect won't it come out as perfect?

Reply:
I might be missing something here but i dont think that is the case.
If there is 2 total tracks, and 1 perfect track 
then tracked_well will be 1 and total_tracks will be 2  = 50%

It is possible if all the new tracks matched to the first expected track that it will be 100% when it is not.